warehouse/schema/schema.go

```diff
@@ -411,20 +229,148 @@ func enhanceSchemaWithIDResolution(consolidatedSchema model.Schema, isIDResoluti
 	return consolidatedSchema
 }
 
-func (sh *Schema) isIDResolutionEnabled() bool {
-	return sh.enableIDResolution && slices.Contains(whutils.IdentityEnabledWarehouses, sh.warehouse.Type)
+func (s *Schema) isIDResolutionEnabled() bool {
+	return s.enableIDResolution && slices.Contains(whutils.IdentityEnabledWarehouses, s.warehouse.Type)
+}
+
+func (s *Schema) UpdateLocalSchemaWithWarehouse(ctx context.Context, uploadID int64) error {
+	s.schemaInWarehouseMu.RLock()
+	defer s.schemaInWarehouseMu.RUnlock()
+	return s.updateLocalSchema(ctx, uploadID, s.schemaInWarehouse)
+}
+
+func (s *Schema) UpdateLocalSchema(ctx context.Context, uploadID int64, updatedSchema model.Schema) error {
+	return s.updateLocalSchema(ctx, uploadID, updatedSchema)
+}
+
+// updateLocalSchema
+// 1. Inserts the updated schema into the local schema table
+// 2. Updates the local schema instance
+func (s *Schema) updateLocalSchema(ctx context.Context, uploadId int64, updatedSchema model.Schema) error {
+	_, err := s.schemaRepo.Insert(ctx, &model.WHSchema{
+		UploadID:        uploadId,
+		SourceID:        s.warehouse.Source.ID,
+		Namespace:       s.warehouse.Namespace,
+		DestinationID:   s.warehouse.Destination.ID,
+		DestinationType: s.warehouse.Type,
+		Schema:          updatedSchema,
+	})
+	if err != nil {
+		return fmt.Errorf("updating local schema: %w", err)
+	}
+
+	s.localSchemaMu.Lock()
+	s.localSchema = updatedSchema
+	s.localSchemaMu.Unlock()
+
+	return nil
+}
+
+// SyncRemoteSchema
+// 1. Fetches schema from local
+// 2. Fetches schema from warehouse
+// 3. Initialize local schema
+// 4. Updates local schema with warehouse schema if it has changed
+// 5. Returns true if schema has changed
+func (s *Schema) SyncRemoteSchema(ctx context.Context, fetchSchemaRepo fetchSchemaRepo, uploadID int64) (bool, error) {
+	localSchema, err := s.GetLocalSchema(ctx)
+	if err != nil {
+		return false, fmt.Errorf("fetching schema from local: %w", err)
+	}
+
+	if err := s.FetchSchemaFromWarehouse(ctx, fetchSchemaRepo); err != nil {
+		return false, fmt.Errorf("fetching schema from warehouse: %w", err)
+	}
+
+	s.localSchemaMu.Lock()
+	s.localSchema = localSchema
+	s.localSchemaMu.Unlock()
```

Comment:
This was missing from earlier code when we made changes around it during [this](https://github.com/rudderlabs/rudder-server/pull/3900).

Reply:
Aren't we updating this if the schema has changed?

```go
	if schemaChanged {
		err := sh.updateLocalSchema(ctx, uploadID, sh.schemaInWarehouse)
		if err != nil {
			return false, fmt.Errorf("updating local schema: %w", err)
		}
	}
```

Reply:
This is only when there is a difference in local and warehouse schema. If not local schema is left empty.

Reply:
I'm not convinced updating it twice is the answer. Let me recheck the `hasSchemaChanged` logic.

Reply:
Currently, I am initializing it with what I got from `sh.GetLocalSchema(ctx)` otherwise it is empty right if there is no difference?

Reply:
Yes, it doesn't work if empty, you are correct :+1: